OpenRA.Mods.Common/Scripting/Global/ActorGlobal.cs

```diff
@@ -64,14 +64,52 @@ public Actor Create(string type, bool addToWorld, LuaTable initTable)
 			return a;
 		}
 
-		[Desc("Returns the build time (in ticks) of the requested unit type.")]
-		public int BuildTime(string type)
+		[Desc("Returns the build time (in ticks) of the requested unit type."
+			+ "An optional second value can be used to exactly specify the producer unit type.")]
+		public int BuildTime(string type, string producer = null)
 		{
 			ActorInfo ai;
 			if (!Context.World.Map.Rules.Actors.TryGetValue(type, out ai))
 				throw new LuaException("Unknown actor type '{0}'".F(type));
 
-			return ai.GetBuildTime();
+			var bi = ai.TraitInfoOrDefault<BuildableInfo>();
+
+			if (bi == null)
+				return 0;
+
+			var time = bi.BuildDuration;
+			if (time == -1)
+			{
+				var valued = ai.TraitInfoOrDefault<ValuedInfo>();
+				time = valued != null ? valued.Cost : 0;
```

Comment:
We should already `return 0` here if `valued == null` (or `valued.Cost == 0`).
That'll save us all the lookups from below.